### PR TITLE
US 016: Create a Get Operation to Obtain All Users

### PR DESCRIPTION
**Pull request by EmmanuelQA**

### CODE REVIEW

File: src/test/java/com/at/globalclasses/ApiTools.java [hunk 1]
```diff
@@ -5,12 +5,17 @@
 import java.io.InputStream;
 import java.util.Properties;
 
+import com.mongodb.client.FindIterable;
```
Comment (ramiroAT) [inline src/test/java/com/at/globalclasses/ApiTools.java]: Why do you add this dependencies?

File: src/test/java/com/at/globalclasses/BasicSecurityUtil.java [hunk 2]
```diff
@@ -6,7 +6,7 @@
     public String environment, requestBody, responseBody, apiResource;
     public ResponseEntity<String> response;
     public ApiTools ServiceApi;
-    public String type;
+    public String type, uridb;
```
Comment (ramiroAT) [inline src/test/java/com/at/globalclasses/BasicSecurityUtil.java:10]: Why did you add this new variable to this class?

File: src/test/java/com/at/globalclasses/MongoDBConnection.java [hunk 3]
```diff
@@ -156,6 +162,72 @@ public boolean compareJsonString(String collection, String object) {
         return bool;
 
     }
+
```
Comment (ramiroAT) [inline src/test/java/com/at/globalclasses/MongoDBConnection.java:163]: Let's move this method to MongoDBUtils class

File: src/test/java/com/at/globalclasses/QAUtils.java [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        } catch(Exception var9) {
+            var9.printStackTrace();
+
+        } finally {
+            db.close();
+        }
+        return bool;
+    }
+
+public long numberOfUsersDB(String env, String mDataBase, String collection) {
```
Comment (ramiroAT) [inline src/test/java/com/at/globalclasses/QAUtils.java:120]: What is the difference between this method and "collectionSize" method from MongoDBUtils class?

File: src/test/java/com/at/globalclasses/QAUtils.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        } catch(Exception var9) {
+            var9.printStackTrace();
+        } finally {
+            db.close();
+        }
+        return number;
+    }
+
+
+////// Encapsulation
```
Comment (ramiroAT) [inline src/test/java/com/at/globalclasses/QAUtils.java]: Why encapsulating and not making only public static methods?

File: src/test/java/com/at/stepdefinitions/Get.java [hunk 6]
```diff
@@ -1,51 +1,142 @@
...
 import io.cucumber.java.en.Then;
 import io.cucumber.java.en.When;
+import org.json.JSONArray;
+import org.json.JSONObject;
+import gherkin.deps.com.google.gson.JsonParser;
 import org.junit.Assert;
 
 public class Get {
     private BasicSecurityUtil base;
+    private BasicSecurityUtil basedb;
```
Comment (ramiroAT) [inline src/test/java/com/at/stepdefinitions/Get.java]: What is the difference between this property and the one above?

File: src/test/java/com/at/stepdefinitions/Get.java [hunk 7]
```diff
@@ -1,51 +1,142 @@
...
+        System.out.println("Users info in the DB: " + firstArray.toString());
+        System.out.println("Users info in the GET op: " + secondArray.toString());
+        QAUtils match = new QAUtils();
+        match.setNumberUsersDB(numberUsers);
+        boolean bool = match.compareDocumentsArrays(firstArray, secondArray);
+        if (bool) {
+            System.out.println("The values of all users match.");
+        } else {
+            System.out.println("The values doesn't match.");
+            throw new io.cucumber.java.PendingException();
```
Comment (ramiroAT) [inline src/test/java/com/at/stepdefinitions/Get.java]: Why throwing a pendingException?

File: src/test/java/com/at/stepdefinitions/Get.java [hunk 8]
```diff
@@ -1,51 +1,142 @@
...
+        ApiTools api = new ApiTools(base.environment, base.apiResource);
+        String usersInfoString = api.retrieve(base.ServiceApi.hostName + base.apiResource).getBody();
+        JSONArray secondArray = new JSONArray(usersInfoString);
+
+        System.out.println("Users info in the DB: " + firstArray.toString());
+        System.out.println("Users info in the GET op: " + secondArray.toString());
+        QAUtils match = new QAUtils();
+        match.setNumberUsersDB(numberUsers);
+        boolean bool = match.compareDocumentsArrays(firstArray, secondArray);
+        if (bool) {
```
Comment (ramiroAT) [inline src/test/java/com/at/stepdefinitions/Get.java]: There is a method in Assert library that can be used to validate a boolean.

File: src/test/java/com/at/stepdefinitions/Get.java [hunk 9]
```diff
@@ -1,51 +1,142 @@
...
+
     }
 
-    @When("I send a GET request")
+    @Given("I want to retrieve all users")  //ready
+    public void i_want_to_retrieve_all_users() throws Exception {
+        base.apiResource = ApiPaths.SCE_GET_USERS;
+    }
+
+    @When("I send a GET request")   //ready
```
Comment (ramiroAT) [inline src/test/java/com/at/stepdefinitions/Get.java]: Please remove comments that are not longer used.

File: src/test/resources/GET.feature [hunk 10]
```diff
@@ -1,15 +1,24 @@
...
+  Scenario: Get all registered users in the DB with the correct information when the number of users is > 0
   Given I want to retrieve all users
   When I send a GET request
   Then the status code should be "200"
-  And information retrieved from service should match with DB
\ No newline at end of file
+  And the number of registered users should be the same of the number of retrieved users
+  And information retrieved from service should match with DB
+  @2
+  Scenario: Get an empty list when there aren't registered users in the DB
```
Comment (ramiroAT) [inline src/test/resources/GET.feature:18]: How are you going to validate this scenario?

**Review by ramiroAT — CHANGES_REQUESTED** [review]
Please take a look to all comments.

File: src/test/java/com/at/globalclasses/ApiTools.java (hunk 1 above)
Comment (EmmanuelQA) [reply src/test/java/com/at/globalclasses/ApiTools.java]: I will delete it. I add this because I thought that i went to use it.

File: src/test/java/com/at/globalclasses/BasicSecurityUtil.java (hunk 2 above)
Comment (EmmanuelQA) [reply src/test/java/com/at/globalclasses/BasicSecurityUtil.java:10]: Because I use these variable to call the DB URL, and the firts version of this code use a similar variable to call the path to make the GET request.
Do I have to change it?

File: src/test/java/com/at/globalclasses/MongoDBConnection.java (hunk 3 above)
Comment (EmmanuelQAstaneda) [reply src/test/java/com/at/globalclasses/MongoDBConnection.java:165]: Ready.
I don't add this method but I already moved it.

Comment (EmmanuelQA) [reply src/test/java/com/at/globalclasses/MongoDBConnection.java:163]: I didn't add this method to the code. I am trying to move it, but I have problems because there are other method with the same name in MONGO DB Utils, but it is only to implement the connection with the DB.

File: src/test/java/com/at/globalclasses/QAUtils.java (hunk 4 above)
Comment (EmmanuelQA) [reply src/test/java/com/at/globalclasses/QAUtils.java:120]: The method "collectionSize" is in MongoDBConnection Class, it obtains the number of documents that are in a collection.
The method "Number of users" implements the connection with the database, and gives to "collection size" the information of the database (URL and COLLECTION) to use.

File: src/test/java/com/at/globalclasses/QAUtils.java (hunk 5 above)
Comment (EmmanuelQA) [reply src/test/java/com/at/globalclasses/QAUtils.java]: Those are just the declarations of "Set" and "Get" of the variable:
private long numbesUsersDB;
I will remove the comment of "encapsulations"

File: src/test/java/com/at/stepdefinitions/Get.java (hunk 6 above)
Comment (EmmanuelQA) [reply src/test/java/com/at/stepdefinitions/Get.java]: "base" take the info to make the GET request.
and "basedb" take the info to implement connection with the DataBase

File: src/test/java/com/at/stepdefinitions/Get.java (hunk 7 above)
Comment (EmmanuelQAstaneda) [reply src/test/java/com/at/stepdefinitions/Get.java:138]: It was to see if the test fail, because above I implement a method that returns a boolean variable, if the boolean is true, print "The values of all users match", if boolean is false aplies  a throwing exception.
I didn't know if there is other way to show if the test fail.

File: src/test/java/com/at/stepdefinitions/Get.java (hunk 8 above)
Comment (EmmanuelQAstaneda) [reply src/test/java/com/at/stepdefinitions/Get.java:134]: ready.

File: src/test/java/com/at/stepdefinitions/Get.java (hunk 9 above)
Comment (EmmanuelQAstaneda) [reply src/test/java/com/at/stepdefinitions/Get.java:46]: ready.

Comment (EmmanuelQA) [reply src/test/java/com/at/stepdefinitions/Get.java]: Done.

File: src/test/java/com/at/stepdefinitions/Get.java (hunk 8 above)
Comment (EmmanuelQA) [reply src/test/java/com/at/stepdefinitions/Get.java]: Done

File: src/test/java/com/at/stepdefinitions/Get.java (hunk 7 above)
Comment (EmmanuelQA) [reply src/test/java/com/at/stepdefinitions/Get.java]: Done.
I used the Assert method.

File: src/test/resources/GET.feature (hunk 10 above)
Comment (EmmanuelQA) [reply src/test/resources/GET.feature:18]: When I tested this scenario, I erased all users of the data base manually.
I have the reports of this scenario.

**Review by ramiroAT — APPROVED** [review]

**Review by IorchAT — APPROVED** [review]
